miner/miner.go

```diff
@@ -47,205 +44,136 @@ type Backend interface {
-	miner.worker.setGasTip(tip)
-	return nil
-}
-
-// SetRecommitInterval sets the interval for sealing work resubmitting.
-func (miner *Miner) SetRecommitInterval(interval time.Duration) {
-	miner.worker.setRecommitInterval(interval)
-}
-
-// Pending returns the currently pending block and associated state. The returned
-// values can be nil in case the pending block is not initialized
-func (miner *Miner) Pending() (*types.Block, *state.StateDB) {
-	return miner.worker.pending()
-}
-
-// PendingBlock returns the currently pending block. The returned block can be
-// nil in case the pending block is not initialized.
-//
-// Note, to access both the pending block and the pending state
-// simultaneously, please use Pending(), as the pending state can
-// change between multiple method calls
-func (miner *Miner) PendingBlock() *types.Block {
-	return miner.worker.pendingBlock()
-}
-
-// PendingBlockAndReceipts returns the currently pending block and corresponding receipts.
-// The returned values can be nil in case the pending block is not initialized.
-func (miner *Miner) PendingBlockAndReceipts() (*types.Block, types.Receipts) {
-	return miner.worker.pendingBlockAndReceipts()
+// Etherbase returns the address of fee recipient.
+func (miner *Miner) Etherbase() common.Address {
+	miner.confMu.RLock()
+	addr := miner.config.Etherbase
+	miner.confMu.RUnlock()
+	return addr
 }
 
+// SetEtherbase sets the address of fee recipient.
 func (miner *Miner) SetEtherbase(addr common.Address) {
-	miner.worker.setEtherbase(addr)
+	miner.confMu.Lock()
+	miner.config.Etherbase = addr
+	miner.confMu.Unlock()
 }
 
 // SetGasCeil sets the gaslimit to strive for when mining blocks post 1559.
 // For pre-1559 blocks, it sets the ceiling.
 func (miner *Miner) SetGasCeil(ceil uint64) {
-	miner.worker.setGasCeil(ceil)
+	miner.confMu.Lock()
+	miner.config.GasCeil = ceil
+	miner.confMu.Unlock()
 }
 
-// SubscribePendingLogs starts delivering logs from pending transactions
-// to the given channel.
-func (miner *Miner) SubscribePendingLogs(ch chan<- []*types.Log) event.Subscription {
-	return miner.worker.pendingLogsFeed.Subscribe(ch)
+// SetGasTip sets the minimum gas tip for inclusion.
+func (miner *Miner) SetGasTip(tip *big.Int) error {
+	miner.confMu.Lock()
+	miner.config.GasPrice = tip
+	miner.confMu.Unlock()
+	return nil
 }
 
 // BuildPayload builds the payload according to the provided parameters.
 func (miner *Miner) BuildPayload(args *BuildPayloadArgs) (*Payload, error) {
-	return miner.worker.buildPayload(args)
+	return miner.buildPayload(args)
+}
+
+// getPending retrieves the pending block based on the current head block.
+// The result might be nil if pending generation is failed.
+func (miner *Miner) getPending() *newPayloadResult {
+	miner.confMu.RLock()
+	coinbase := miner.config.Etherbase
+	miner.confMu.RUnlock()
+
+	header := miner.chain.CurrentHeader()
+	if cached := miner.pending.resolve(header.Hash(), coinbase); cached != nil {
+		return cached
+	}
```

Comment:
My nit with miner.Pending() and/or miner.getPending() is that if a burst of requests are made (which is expected IMO), all the requests will cause a new Pending block to be built.

We should have a mean to track that a pending block is currently being built and have subsequent requests wait it out vs doing concurrent builds.

Reply:
A brute force way would be

```
miner.pendMu.Lock()
if cached := miner.pending.resolve(header.Hash(), coinbase); cached != nil {
    miner.pendMu.Unlock()
    return cached
}
defer miner.pendMu.Unlock()
```

Reply:
Actually, it's a fairly elegant way IMO and self-contained.